File: utils/git/actions.py
import subprocess
from rich.panel import Panel
from rich import box
from typing import Optional

class GitActions:
    """
    Primary Git workflows: commit, push, branch management, and reset.
    """
    # This class expects run_command, get_current_branch, and console to be available 
    # (via GitCore mixin).

    def commit_and_push(self, message: str, path: str = ".", no_verify: bool = False) -> bool:
        """
        Stage changes, commit, and push. Automatically handles upstream tracking.
        """
        with self.console.status("[bold green]Working on your commit...[/]"):
            if not self.run_command(["add", path]): return False
            
            commit_cmd = ["commit", "-m", message]
            if no_verify:
                commit_cmd.append("--no-verify")
                
            if not self.run_command(commit_cmd): return False
            
            branch = self.get_current_branch()
            has_upstream = False
            if branch:
                try:
                    subprocess.check_output(["git", "rev-parse", "--symbolic-full-name", "@{u}"], stderr=subprocess.DEVNULL)
                    has_upstream = True
                except subprocess.CalledProcessError:
                    has_upstream = False

            if has_upstream:
                push_success = self.run_command(["push"])
            else:
                push_success = self.run_command(["push", "--set-upstream", "origin", branch])

            if not push_success: return False
        
        self.console.print(Panel(
            f"Message: [bold white]{message}[/]\nStatus: [bold green]Pushed to Remote[/]",
            title="Commit & Push", border_style="green", box=box.ROUNDED
        ))
        return True

    def reset_commit(self, mode: str = "soft") -> bool:
        """
        Reset the current branch head to the previous commit.
        """
        if mode not in ["soft", "hard"]:
            self.console.print("[bold red]Error:[/] Invalid mode.")
            return False
            
        with self.console.status(f"[bold red]Resetting: {mode}...[/]"):
            success = self.run_command(["reset", f"--{mode}", "HEAD~1"])
        
        if success:
            self.console.print(Panel(
                f"Reset to [bold]HEAD~1[/] using [bold red]{mode}[/] mode.",
                title="Git Reset", border_style="red", box=box.ROUNDED
            ))
        return success

    def switch_branch(self, name: str) -> bool:
        """
        Create a new branch and switch to it immediately.
        """
        with self.console.status(f"[bold blue]Switching to {name}...[/]"):
            if not self.run_command(["checkout", "-b", name]):
                return False
            success = self.run_command(["push", "-u", "origin", name])
            
        if success:
            self.console.print(Panel(
                f"Active Branch: [bold green]{name}[/]\nTracking origin.",
                title="Branch Switch", border_style="blue", box=box.ROUNDED
            ))
        return success

File: internal/git/actions.go
package git

import (
	"errors"
	"fmt"
	"strings"
)

type GitActions struct {
	Core *GitCore
}

func NewGitActions(core *GitCore) (*GitActions, error) {
	if core == nil {
		return nil, errors.New("git core is required")
	}
	return &GitActions{Core: core}, nil
}

func (g *GitActions) AddAll() error {
	if _, stderr, err := g.Core.RunCommand("add", "."); err != nil {
		if stderr != "" {
			return fmt.Errorf("git add failed: %s", stderr)
		}
		return fmt.Errorf("git add failed: %w", err)
	}
	return nil
}

func (g *GitActions) CommitAndPush(message string, noVerify bool) error {
	if strings.TrimSpace(message) == "" {
		return errors.New("commit message cannot be empty")
	}

	commitArgs := []string{"commit", "-m", message}
	if noVerify {
		commitArgs = append(commitArgs, "--no-verify")
	}
	if _, stderr, err := g.Core.RunCommand(commitArgs...); err != nil {
		if stderr != "" {
			return fmt.Errorf("git commit failed: %s", stderr)
		}
		return fmt.Errorf("git commit failed: %w", err)
	}

	return g.Push(noVerify)
}

func (g *GitActions) Push(noVerify bool) error {
	hasUpstream, err := g.hasUpstream()
	if err != nil {
		return err
	}

	if hasUpstream {
		pushArgs := []string{"push"}
		if noVerify {
			pushArgs = append(pushArgs, "--no-verify")
		}
		if _, stderr, err := g.Core.RunCommand(pushArgs...); err != nil {
			if stderr != "" {
				return fmt.Errorf("git push failed: %s", stderr)
			}
			return fmt.Errorf("git push failed: %w", err)
		}
		return nil
	}

	branch, err := g.Core.GetCurrentBranch()
	if err != nil {
		return fmt.Errorf("failed to detect branch for upstream push: %w", err)
	}

	pushArgs := []string{"push"}
	if noVerify {
		pushArgs = append(pushArgs, "--no-verify")
	}
	pushArgs = append(pushArgs, "--set-upstream", "origin", branch)
	if _, stderr, err := g.Core.RunCommand(pushArgs...); err != nil {
		if stderr != "" {
			return fmt.Errorf("git push --set-upstream failed: %s", stderr)
		}
		return fmt.Errorf("git push --set-upstream failed: %w", err)
	}

	return nil
}

func (g *GitActions) CreateBranch(name string) error {
	name = strings.TrimSpace(name)
	if name == "" {
		return errors.New("branch name cannot be empty")
	}
	if _, stderr, err := g.Core.RunCommand("switch", "-c", name); err != nil {
		if stderr != "" {
			return fmt.Errorf("git switch -c %s failed: %s", name, stderr)
		}
		return fmt.Errorf("git switch -c %s failed: %w", name, err)
	}
	if _, stderr, err := g.Core.RunCommand("push", "-u", "origin", name); err != nil {
		if stderr != "" {
			return fmt.Errorf("git push -u origin %s failed: %s", name, stderr)
		}
		return fmt.Errorf("git push -u origin %s failed: %w", name, err)
	}
	return nil
}

func (g *GitActions) SwitchBranch(name string) error {
	name = strings.TrimSpace(name)
	if name == "" {
		return errors.New("branch name cannot be empty")
	}
	if _, stderr, err := g.Core.RunCommand("switch", name); err != nil {
		if stderr != "" {
			return fmt.Errorf("git switch %s failed: %s", name, stderr)
		}
		return fmt.Errorf("git switch %s failed: %w", name, err)
	}
	return nil
}

func (g *GitActions) HasUpstream() (bool, error) {
	return g.hasUpstream()
}

func (g *GitActions) hasUpstream() (bool, error) {
	_, stderr, err := g.Core.RunCommand("rev-parse", "--symbolic-full-name", "@{u}")
	if err == nil {
		return true, nil
	}

	low := strings.ToLower(stderr)
	if strings.Contains(low, "no upstream configured") ||
		strings.Contains(low, "does not have any upstream branch") {
		return false, nil
	}

	if stderr != "" {
		return false, fmt.Errorf("failed checking upstream branch: %s", stderr)
	}
	return false, fmt.Errorf("failed checking upstream branch: %w", err)
}
